src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,411 @@
+    private Boolean searchSortPartOfIndexSort = null;
+
+    private static final TopFieldDocs EMPTY_TOPDOCS = new TopFieldDocs(
+        new TotalHits(0, TotalHits.Relation.EQUAL_TO),
+        new ScoreDoc[0],
+        new SortField[0]
+    );
+
+    // Declaring the constructor private prevents extending this class by anyone
+    // else. Note that the class cannot be final since it's extended by the
+    // internal versions. If someone will define a constructor with any other
+    // visibility, then anyone will be able to extend the class, which is not what
+    // we want.
+    public HybridTopFieldDocSortCollector(final int numHits, final HitsThresholdChecker hitsThresholdChecker) {
+        this.numHits = numHits;
+        this.hitsThresholdChecker = hitsThresholdChecker;
+    }
+
+    // Add the entry in the Priority queue
+    private void add(int slot, int doc, FieldValueHitQueue<FieldValueHitQueue.Entry> compoundScore, int subQueryNumber, float score) {
+        FieldValueHitQueue.Entry bottomEntry = new FieldValueHitQueue.Entry(slot, docBase + doc);
+        bottomEntry.score = score;
+        bottom = compoundScore.add(bottomEntry);
+        // The queue is full either when totalHits == numHits (in SimpleFieldCollector), in which case
+        // slot = totalHits - 1, or when hitsCollected == numHits (in PagingFieldCollector this is hits
+        // on the current page) and slot = hitsCollected - 1.
+        assert slot < numHits;
+        queueFull[subQueryNumber] = slot == numHits - 1;
+    }
+
+    private void updateBottom(int doc, FieldValueHitQueue<FieldValueHitQueue.Entry> compoundScore) {
+        bottom.doc = docBase + doc;
+        bottom = compoundScore.updateTop();
+    }
+
+    private boolean canEarlyTerminate(Sort searchSort, Sort indexSort) {
+        return canEarlyTerminateOnDocId(searchSort) || canEarlyTerminateOnPrefix(searchSort, indexSort);
+    }
+
+    private boolean canEarlyTerminateOnDocId(Sort searchSort) {
+        final SortField[] fields1 = searchSort.getSort();
+        return SortField.FIELD_DOC.equals(fields1[0]);
+    }
+
+    private boolean canEarlyTerminateOnPrefix(Sort searchSort, Sort indexSort) {
+        if (indexSort != null) {
+            final SortField[] fields1 = searchSort.getSort();
+            final SortField[] fields2 = indexSort.getSort();
+            // early termination is possible if fields1 is a prefix of fields2
+            if (fields1.length > fields2.length) {
+                return false;
+            }
+            return Arrays.asList(fields1).equals(Arrays.asList(fields2).subList(0, fields1.length));
+        } else {
+            return false;
+        }
+    }
+
+    public List<TopFieldDocs> topDocs(final FieldValueHitQueue<FieldValueHitQueue.Entry>[] compoundScores, final Sort sort) {
+        if (compoundScores == null) {
+            return new ArrayList<>();
+        }
+        final List<TopFieldDocs> topFieldDocs = IntStream.range(0, compoundScores.length)
```

Comment:
given the impact we have seen of using streams on performance I would make this as a simple for loop. To ensure that we are not doing another round of optimizations on these streams.